commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/hull/ConvexHull3D.java

```diff
@@ -0,0 +1,697 @@
+        region = null;
+        this.facets = Collections.emptySet();
+        this.isDegenerate = isDegenerate;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public List<Vector3D> getVertices() {
+        return vertices;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public ConvexVolume getRegion() {
+        return region;
+    }
+
+    /**
+     * Return a collection of all two-dimensional faces (called facets) of the
+     * convex hull.
+     *
+     * @return a collection of all two-dimensional faces.
+     */
+    public Collection<? extends ConvexPolygon3D> getFacets() {
+        return Collections.unmodifiableCollection(facets);
+    }
+
+    /**
+     * Return {@code true} if the hull is degenerate.
+     *
+     * @return the isDegenerate
+     */
+    public boolean isDegenerate() {
+        return isDegenerate;
+    }
+
+    /**
+     * Implementation of quick-hull algorithm by Barber, Dobkin and Huhdanpaa. The
+     * algorithm constructs the convex hull for a given finite set of points.
+     * Empirically, the number of points processed by Quickhull is proportional to
+     * the number of vertices in the output. The algorithm runs on an input of size
+     * n with r processed points in time O(n log r). We define a point of the given
+     * set to be extreme, if and only if the point is part of the final hull. The
+     * algorithm runs in multiple stages:
+     * <ol>
+     * <li>First we construct a simplex with extreme properties from the given point
+     * set to maximize the possibility of choosing extreme points as initial simplex
+     * vertices.</li>
+     * <li>We partition all the remaining points into outside sets. Each polygon
+     * face of the simplex defines a positive and negative half-space. A point can
+     * be assigned to the outside set of the polygon if it is an element of the
+     * positive half space.</li>
+     * <li>For each polygon-face (facet) with a non empty outside set we choose a
+     * point with maximal distance to the given facet.</li>
+     * <li>We determine all the visible facets from the given outside point and find
+     * a path around the horizon.</li>
+     * <li>We construct a new cone of polygons from the edges of the horizon to the
+     * outside point. All visible facets are removed and the points in the outside
+     * sets of the visible facets are redistributed.</li>
+     * <li>We repeat step 3-5 until each outside set is empty.</li>
+     * </ol>
+     */
+    public static class Builder {
+
+        /** Set of possible candidates. */
+        private final PointSet<Vector3D> candidates;
+
+        /** Precision context used to compare floating point numbers. */
+        private final DoubleEquivalence precision;
+
+        /** Simplex for testing new points and starting the algorithm. */
+        private Simplex simplex;
+
+        /** The minX, maxX, minY, maxY, minZ, maxZ points. */
+        private final Vector3D[] box;
```

Comment:
I used bounds although this will slightly altered the geometric structure in which we look for changes. The initial structure was more like a double pyramid.

Reply:
Can you explain this geometric difference a bit more?